File: 56_merge_intervals/merge.go
func merge(intervals [][]int) [][]int {
	var res [][]int

	sort.Slice(intervals, func(i, j int) bool {
		if len(intervals[i]) == 0 && len(intervals[j]) == 0 {
			return false
		}
		if len(intervals[i]) == 0 || len(intervals[j]) == 0 {
			return len(intervals[i]) == 0
		}

		return intervals[i][0] < intervals[j][0]
	})

	res = append(res, intervals[0])

	for _, interval := range intervals {
		n := len(res) - 1
		if res[n][1] >= interval[0] {
			res[n][1] = max(res[n][1], interval[1])
		} else {
			res = append(res, interval)
		}
	}

	return res
}

func max(a int, b int) int {
	if a > b {
		return a
	}

	return b
}

File: 56_merge_intervals/merge.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        vector<vector<int>> res;
        sort(intervals.begin(), intervals.end());

        for (auto interval : intervals) {
            if (res.empty() || res.back()[1] < interval[0]) {
                res.push_back(interval);
            } else {
                res.back()[1] = max(res.back()[1], interval[1]);
            }
        }

        return res;
    }

    static bool comp(vector<int> a, vector<int> b) { return a[0] > b[0]; }
};